router.go

```diff
@@ -63,39 +72,76 @@ func (m MapRouter) Register(rs []transport.Procedure) {
 			panic("Expected procedure name not to be empty string in registration")
 		}
 
+		if r.Encoding != "" {
+			// Route to individual handlers for unique combinations of service,
+			// procedure, and encoding. This shall henceforth be the
+			// recommended way for models to register procedures.
+			spe := serviceProcedureEncoding{
+				service:   r.Service,
+				procedure: r.Name,
+				encoding:  r.Encoding,
+			}
+			m.serviceProcedureEncodings[spe] = r
+		}
+
+		// This supports wild card encodings (for backward compatibility,
+		// since type models like Thrift were not previously required to
+		// specify the encoding of every procedure)
 		sp := serviceProcedure{
 			service:   r.Service,
 			procedure: r.Name,
 		}
-		m.entries[sp] = r
+		// Protect against registering ambiguous wildcards
+		if _, ok := m.serviceProcedures[sp]; r.Encoding == "" && ok {
```

Comment:
Do you also want to check if this is in `m.serviceProcedureEncodings`? I think we should outlaw registering a `transport.Procedure` with an encoding, and without an encoding - there should either be one `transport.Procedure` without an encoding, or >=1 `transport.Procedure`s with an encoding.